// uu_appg01_devkit/src/config/eslint.config.js
const js = require("@eslint/js");
const prettier = require("eslint-plugin-prettier");
const json = require("eslint-plugin-json");
const globals = require("globals");

module.exports = [
  {
    files: ["**/*.js"],

    plugins: {
      prettier,
    },

    languageOptions: {
      globals: {
        ...globals.commonjs,
        ...globals.node,
        ...globals.jest,
      },

      ecmaVersion: 2023,
      sourceType: "commonjs",
    },

    rules: {
      ...js.configs.recommended.rules,

      "no-const-assign": "warn",
      "no-this-before-super": "warn",
      "no-undef": "warn",
      "no-unreachable": "warn",

      "no-unused-vars": [
        "warn",
        {
          args: "none",
          ignoreRestSiblings: true,
        },
      ],

      "no-use-before-define": [
        "warn",
        {
          functions: false,
        },
      ],

      "constructor-super": "warn",
      "valid-typeof": "warn",
      "no-console": "off",

      "prettier/prettier": "warn",
      "arrow-body-style": "off",
      "prefer-arrow-callback": "off",
    },
  },
  {
    files: ["**/*.json"],
    ...json.configs.recommended,
  },
];
