// uu_appg01_devkit-common/src/templates/__nodejs-app/app/constants/project-constants.js
"use strict";

//@@viewOn:constants
const <%=  context.projectClassName %>Constants = {
  AWSC_PREFIX: "<%=  context.namespaceAppSnakeCase %>",
  CONSOLE_PREFIX: "<%=  context.app %>",
  ERROR_PREFIX: "<%=  context.errorCode %>",
  INIT_PROGRESS_CODE_PREFIX: "<%=  context.insomniaContextPath %>-progress-init-",
  INIT_PROGRESS_NAME_PREFIX: "<%=  context.namespaceAppCamelCase %> Init ",
  SET_STATE_CLOSED_PROGRESS_CODE_PREFIX: "<%=  context.insomniaContextPath %>-progress-setStateClosed-",
  SET_STATE_CLOSED_PROGRESS_NAME_PREFIX: "<%=  context.namespaceAppCamelCase %> Set State Closed ",
  CLEAR_PROGRESS_CODE_PREFIX: "<%=  context.insomniaContextPath %>-progress-clear-",
  CLEAR_PROGRESS_NAME_PREFIX: "<%=  context.namespaceAppCamelCase %> Clear ",
  UUAPP_CODE: "<%=  context.insomniaContextPath %>",

  CONFIG_CACHE_KEY: "configuration",
  UU_APP_ERROR_MAP: "uuAppErrorMap",

  // This is bound matrix of uuAwsc and uuConsole which has authorization bounded to that uuAwsc.
  CONSOLE_BOUND_MATRIX: {
    Authorities: ["Authorities", "Readers", "Writers"],
    Operatives: ["Readers", "Writers"],
    Auditors: ["Readers"],
    SystemIdentity: ["Authorities", "Readers", "Writers"],
  },

  InitStepMap: {
    <%=  context.app.toUpperCase() %>_OBJECT_CREATED: { code: "<%=  context.namespaceAppCamelCase %>ObjectCreated", message: "The uuObject of <%=  context.namespaceAppCamelCase %> created." },
    AWSC_CREATED: { code: "uuAwscCreated", message: "The uuAwsc of <%=  context.namespaceAppCamelCase %> created." },
    WS_CONNECTED: { code: "uuAppWorkspaceConnected", message: "The <%=  context.namespaceAppCamelCase %> uuAppWorkspace connected." },
    CONSOLE_CREATED: { code: "consoleCreated", message: "The console of <%=  context.namespaceAppCamelCase %> created." },
    PROGRESS_ENDED: { code: "progressEnded", message: "The progress has been ended." },
    WS_ACTIVE: { code: "uuAppWorkspaceActiveState", message: "The uuAppWorkspace of <%=  context.namespaceAppCamelCase %> set to active state." },
  },

  InitRollbackStepMap: {
    CONSOLE_CLEARED: { code: "consoleCleared", message: "The <%=  context.namespaceAppCamelCase %> console has been cleared." },
    WS_DISCONNECTED: { code: "uuAppWorkspaceDisonnected", message: "The <%=  context.namespaceAppCamelCase %> uuAppWorkspace disconnected." },
    AWSC_DELETED: { code: "uuAwscDeleted", message: "The uuAwsc of <%=  context.namespaceAppCamelCase %> deleted." },
    PROGRESS_DELETED: { code: "progressDeleted", message: "The progress has been deleted." },
  },

  SetStateClosedStepMap: {
    CLOSE_STARTED: { code: "setStateClosedStarted", message: "SetStateClosed has started." },
    AWSC_CLOSED: { code: "uuAwscClosed", message: "The uuObject of <%=  context.namespaceAppCamelCase %> set to closed state." },
    PROGRESS_ENDED: { code: "progressEnded", message: "The progress has been ended." },
  },

  ClearStepMap: {
    CLEAR_STARTED: { code: "clearStarted", message: "Clear has started." },
    INIT_PROGRESS_DELETED: { code: "initProgressDeleted", message: "The init progress has been deleted." },
    SET_STATE_CLOSED_PROGRESS_DELETED: {
      code: "setStateClosedProgressDeleted",
      message: "The setStateClosed progress has been deleted.",
    },
    CONSOLE_CLEARED: { code: "consoleCleared", message: "The <%=  context.namespaceAppCamelCase %> console has been cleared." },
    PROGRESS_AUTH_STRATEGY_SET: {
      code: "progressAuthorizationStrategySet",
      message: "The authorization strategy of progress has been set to roleGroupInterface.",
    },
    AUTH_STRATEGY_SET: {
      code: "authorizationStrategySet",
      message: "The authorization strategy has been set to roleGroupInterface.",
    },
    AWSC_DELETED: { code: "uuAwscDeleted", message: "The uuAwsc of <%=  context.namespaceAppCamelCase %> deleted." },
    PROGRESS_ENDED: { code: "progressEnded", message: "The progress has been ended." },
  },

  ModeMap: {
    STANDARD: "standard",
    RETRY: "retry",
    ROLLBACK: "rollback",
  },

  ProfileMask: {
    STANDARD_USER: parseInt("00010000000000000000000000000000", 2),
  },

  PropertyMap: {
    CONFIG: "config",
    SCRIPT_CONFIG: "scriptConfig",
    <%=  context.app.toUpperCase() %>_CONFIG: "<%=  context.namespaceAppCamelCase %>Config",
  },

  Schemas: {
    <%=  context.app.toUpperCase() %>_INSTANCE: "<%=  context.projectClassName.charAt(0).toLowerCase() + context.projectClassName.substr(1) %>",
  },

  SharedResources: {
    SCRIPT_CONSOLE: "uu-console-maing02",
    SCRIPT_ENGINE: "uu-script-engineg02",
  },

  StateMap: {
    CREATED: "created",
    BEING_INITIALIZED: "beingInitialized",
    ACTIVE: "active",
    FINAL: "closed",
  },

  getMainConsoleCode: (awid) => {
    return `<%=  context.insomniaContextPath %>-console-${awid}`;
  },

  getInitProgressCode: (awid) => {
    return `${<%=  context.projectClassName %>Constants.INIT_PROGRESS_CODE_PREFIX}${awid}`;
  },

  getInitProgressName: (awid) => {
    return `${<%=  context.projectClassName %>Constants.INIT_PROGRESS_NAME_PREFIX}${awid}`;
  },

  getSetStateClosedProgressName: (awid) => {
    return `${<%=  context.projectClassName %>Constants.SET_STATE_CLOSED_PROGRESS_NAME_PREFIX}${awid}`;
  },

  getSetStateClosedProgressCode: (awid) => {
    return `${<%=  context.projectClassName %>Constants.SET_STATE_CLOSED_PROGRESS_CODE_PREFIX}${awid}`;
  },

  getClearProgressName: (awid) => {
    return `${<%=  context.projectClassName %>Constants.CLEAR_PROGRESS_NAME_PREFIX}${awid}`;
  },

  getClearProgressCode: (awid) => {
    return `${<%=  context.projectClassName %>Constants.CLEAR_PROGRESS_CODE_PREFIX}${awid}`;
  },

  getInitStepCount: () => {
    return Object.keys(<%=  context.projectClassName %>Constants.InitStepMap).length;
  },

  getInitRollbackStepCount: () => {
    return Object.keys(<%=  context.projectClassName %>Constants.InitRollbackStepMap).length;
  },

  getSetStateClosedStepCount: () => {
    return Object.keys(<%=  context.projectClassName %>Constants.SetStateClosedStepMap).length;
  },

  getClearStepCount: () => {
    return Object.keys(<%=  context.projectClassName %>Constants.ClearStepMap).length;
  },
};
//@@viewOff:constants

//@@viewOn:exports
module.exports = <%=  context.projectClassName %>Constants;
//@@viewOff:exports
