// uu_appg01_devkit-common/src/templates/__nodejs-app/app/abl/project/clear-abl.js
"use strict";
const Crypto = require("crypto");
const { DaoFactory } = require("uu_appg01_server").ObjectStore;
const { UuAppWorkspace } = require("uu_appg01_server").Workspace;
const { AuthenticationService } = require("uu_appg01_server").Authentication;
const { UriBuilder } = require("uu_appg01_server").Uri;
const { UuDateTime } = require("uu_i18ng01");
const { ConsoleClient, ProgressClient } = require("uu_consoleg02-uulib");

const Errors = require("../../api/errors/<%= context.projectClassFile %>-error");
const Warnings = require("../../api/warnings/<%= context.projectClassFile %>-warning");
const Validator = require("../../components/validator");
const DtoBuilder = require("../../components/dto-builder");
const ScriptEngineClient = require("../../components/script-engine-client");
const <%= context.projectClassName %>Client = require("../../components/<%= context.projectClassFile %>-client");
const StepHandler = require("../../components/step-handler");

const ProgressConstants = require("../../constants/progress-constants");
const <%= context.projectClassName %>Constants = require("../../constants/<%= context.projectClassFile %>-constants");
const Configuration = require("../../components/configuration");

const SCRIPT_CODE = "<%=  context.nameWithVendor.replace(/-server$/, "-uuscriptlib") %>/<%= context.projectClassFile%>/clear";

class ClearAbl {
  constructor() {
    this.dao = DaoFactory.getDao(<%= context.projectClassName %>Constants.Schemas.<%=  context.app.toUpperCase() %>_INSTANCE);
  }

  async clear(uri, dtoIn, session) {
    // HDS 1
    const awid = uri.getAwid();
    Validator.validateDtoIn(uri, dtoIn);

    // HDS 2
    let <%= context.namespaceAppCamelCase %> = await this.dao.getByAwid(awid);

    if (<%= context.namespaceAppCamelCase %>) {
      if (<%= context.namespaceAppCamelCase %>.state !== <%= context.projectClassName %>Constants.StateMap.FINAL) {
        // 2.1
        throw new Errors.Clear.NotInProperState({
          state: <%= context.namespaceAppCamelCase %>.state,
          expectedStateList: [<%= context.projectClassName %>Constants.StateMap.FINAL],
        });
      }

      if (<%= context.namespaceAppCamelCase %>.temporaryData && <%= context.namespaceAppCamelCase %>.temporaryData.useCase !== uri.getUseCase()) {
        // 2.2
        throw new Errors.SetStateClosed.UseCaseExecutionForbidden({
          concurrencyUseCase: <%= context.namespaceAppCamelCase %>.temporaryData.useCase,
        });
      }
    } else {
      try {
        await UuAppWorkspace.setAssignedSysState(awid);
      } catch (e) {
        // 2.3
        throw new Errors.Clear.SetAssignedSysStateFailed({}, e);
      }

      return DtoBuilder.prepareDtoOut({ progressMap: {} });
    }

    // HDS 3
    const configuration = await Configuration.getUuSubAppConfiguration({
      awid,
      artifactId: <%= context.namespaceAppCamelCase %>.artifactId,
      uuTerritoryBaseUri: <%= context.namespaceAppCamelCase %>.uuTerritoryBaseUri,
    });

    // HDS 4
    const sysIdentitySession = await AuthenticationService.authenticateSystemIdentity();
    const lockSecret = Crypto.randomBytes(32).toString("hex");
    const progressClient = await this._createClearProgress(
      <%= context.namespaceAppCamelCase %>,
      dtoIn,
      configuration,
      lockSecret,
      sysIdentitySession,
    );

    // HDS 5
    if (!<%= context.namespaceAppCamelCase %>.temporaryData) {
      <%= context.namespaceAppCamelCase %> = await this.dao.updateByAwid({
        awid,
        temporaryData: {
          useCase: uri.getUseCase(),
          dtoIn: dtoIn.data,
          stepList: [<%= context.projectClassName %>Constants.ClearStepMap.CLEAR_STARTED.code],
          progressMap: {
            progressCode: progressClient.progress.code,
            uuConsoleUri: configuration.uuConsoleBaseUri,
            consoleCode: <%= context.projectClassName %>Constants.getMainConsoleCode(awid),
          },
        },
      });
    }

    if (<%= context.namespaceAppCamelCase %>.temporaryData.stepList.includes(<%= context.projectClassName %>Constants.ClearStepMap.CONSOLE_CLEARED.code)) {
      await this._clearFinalize(uri, { lockSecret }, session);
    } else {
      // TODO If your application requires any additional steps, add them here...
  
      // HDS 6
      await this._runScript(
        uri.getBaseUri(),
        dtoIn,
        configuration,
        progressClient.progress.lockSecret,
        sysIdentitySession,
      );
    }

    // HDS 7
    return DtoBuilder.prepareDtoOut({ data: <%= context.namespaceAppCamelCase %> });
  }

  async _clearFinalize(uri, dtoIn, session) {
    // HDS 1
    const awid = uri.getAwid();
    Validator.validateDtoInCustom(uri, dtoIn, "sysUuAppWorkspaceInitFinalizeDtoInType");

    // HDS 2
    let <%= context.namespaceAppCamelCase %> = await this.dao.getByAwid(awid);

    if (!<%= context.namespaceAppCamelCase %>) {
      // 2.1
      throw new Errors._clearFinalize.<%= context.namespaceAppCamelCase.charAt(0).toUpperCase() + context.namespaceAppCamelCase.slice(1) %>DoesNotExist({ awid });
    }

    if (<%= context.namespaceAppCamelCase %>.state !== <%= context.projectClassName %>Constants.StateMap.FINAL) {
      // 2.2
      throw new Errors._clearFinalize.NotInProperState({
        state: <%= context.namespaceAppCamelCase %>.state,
        expectedStateList: [<%= context.projectClassName %>Constants.StateMap.FINAL],
      });
    }

    if (!<%= context.namespaceAppCamelCase %>.temporaryData) {
      // 2.3
      throw new Errors._clearFinalize.MissingRequiredData();
    }

    if (<%= context.namespaceAppCamelCase %>.temporaryData && <%= context.namespaceAppCamelCase %>.temporaryData.useCase !== "sys/uuAppWorkspace/clear") {
      // 2.4
      throw new Errors._clearFinalize.UseCaseExecutionForbidden({
        concurrencyUseCase: <%= context.namespaceAppCamelCase %>.temporaryData.useCase,
      });
    }

    // HDS 3
    const sysIdentitySession = await AuthenticationService.authenticateSystemIdentity();
    const progress = {
      code: <%= context.projectClassName %>Constants.getClearProgressCode(<%= context.namespaceAppCamelCase %>.awid),
      lockSecret: dtoIn.lockSecret,
    };
    let progressClient = null;
    if (!<%= context.namespaceAppCamelCase %>.temporaryData.stepList.includes(<%= context.projectClassName %>Constants.ClearStepMap.PROGRESS_ENDED.code)) {
      progressClient = await ProgressClient.get(<%= context.namespaceAppCamelCase %>.temporaryData.progressMap.uuConsoleUri, progress, {
        session: sysIdentitySession,
      });
    }
    const stepHandler = new StepHandler({
      schema: <%= context.projectClassName %>Constants.Schemas.<%=  context.app.toUpperCase() %>_INSTANCE,
      progressClient,
      stepList: <%= context.namespaceAppCamelCase %>.temporaryData.stepList,
    });

    // TODO If your application requires any additional steps, add them here...

    // HDS 5
    <%= context.namespaceAppCamelCase %> = await stepHandler.handleStep(<%= context.namespaceAppCamelCase %>, <%= context.projectClassName %>Constants.ClearStepMap.INIT_PROGRESS_DELETED, async () => {
      await this._deleteProgress(
        <%= context.projectClassName %>Constants.getInitProgressCode(awid),
        <%= context.namespaceAppCamelCase %>.temporaryData.progressMap.uuConsoleUri,
        sysIdentitySession,
      );
    });

    // HDS 6
    <%= context.namespaceAppCamelCase %> = await stepHandler.handleStep(
      <%= context.namespaceAppCamelCase %>,
      <%= context.projectClassName %>Constants.ClearStepMap.SET_STATE_CLOSED_PROGRESS_DELETED,
      async () => {
        await this._deleteProgress(
          <%= context.projectClassName %>Constants.getSetStateClosedProgressCode(awid),
          <%= context.namespaceAppCamelCase %>.temporaryData.progressMap.uuConsoleUri,
          sysIdentitySession,
        );
      },
    );

    // HDS 7
    <%= context.namespaceAppCamelCase %> = await stepHandler.handleStep(<%= context.namespaceAppCamelCase %>, <%= context.projectClassName %>Constants.ClearStepMap.CONSOLE_CLEARED, async () => {
      await this._clearConsole(
        <%= context.namespaceAppCamelCase %>.temporaryData.progressMap.uuConsoleUri,
        <%= context.projectClassName %>Constants.getMainConsoleCode(awid),
        sysIdentitySession,
      );
    });

    // HDS 8
    <%= context.namespaceAppCamelCase %> = await stepHandler.handleStep(<%= context.namespaceAppCamelCase %>, <%= context.projectClassName %>Constants.ClearStepMap.PROGRESS_AUTH_STRATEGY_SET, async () => {
      await this._setClearProgressAuthorizationStrategy(
        <%= context.namespaceAppCamelCase %>.temporaryData.progressMap.uuConsoleUri,
        <%= context.namespaceAppCamelCase %>.temporaryData.progressMap.progressCode,
        awid,
        <%= context.namespaceAppCamelCase %>.temporaryData.dtoIn.awidInitiatorList,
        sysIdentitySession,
      );
    });

    // HDS 9
    <%= context.namespaceAppCamelCase %> = await stepHandler.handleStep(<%= context.namespaceAppCamelCase %>, <%= context.projectClassName %>Constants.ClearStepMap.AUTH_STRATEGY_SET, async () => {
      await UuAppWorkspace.setAuthorizationStrategy(
        uri,
        {
          authorizationStrategy: "roleGroupInterface",
          roleGroupUriMap: {},
        },
        session,
      );
    });

    // HDS 10
    <%= context.namespaceAppCamelCase %> = await stepHandler.handleStep(<%= context.namespaceAppCamelCase %>, <%= context.projectClassName %>Constants.ClearStepMap.AWSC_DELETED, async () => {
      const <%= context.projectClassName.charAt(0).toLowerCase() + context.projectClassName.substr(1) %>Client = new <%= context.projectClassName %>Client(<%= context.namespaceAppCamelCase %>, <%= context.namespaceAppCamelCase %>.uuTerritoryBaseUri);
      await <%= context.projectClassName.charAt(0).toLowerCase() + context.projectClassName.substr(1) %>Client.deleteAwsc();
    });

    // HDS 11
    <%= context.namespaceAppCamelCase %> = await stepHandler.handleStep(
      <%= context.namespaceAppCamelCase %>,
      <%= context.projectClassName %>Constants.ClearStepMap.PROGRESS_ENDED,
      async () => {
        await progressClient.end({
          state: ProgressConstants.StateMap.COMPLETED,
          message: "Clear finished.",
          expireAt: UuDateTime.now().shiftDay(1),
          doneWork: <%= context.projectClassName %>Constants.getSetStateClosedStepCount(),
        });
      },
      false,
    );

    // HDS 12
    if (<%= context.namespaceAppCamelCase %>.temporaryData.dtoIn.awidInitiatorList) {
      await UuAppWorkspace.reassign({
        awid,
        awidInitiatorList: <%= context.namespaceAppCamelCase %>.temporaryData.dtoIn.awidInitiatorList,
      });
    }

    // HDS 13
    await this.dao.deleteByAwid(awid);

    // HDS 14
    try {
      await UuAppWorkspace.setAssignedSysState(awid);
    } catch (e) {
      throw new Errors._clearFinalize.SetAssignedSysStateFailed({}, e);
    }

    // HDS 15
    return DtoBuilder.prepareDtoOut();
  }

  _parseTerritoryUri(locationUri) {
    let uuTerritoryUri;

    try {
      uuTerritoryUri = UriBuilder.parse(locationUri);
    } catch (e) {
      throw new Errors.Clear.UuTLocationUriParseFailed({ uri: locationUri }, e);
    }

    return uuTerritoryUri;
  }

  async _createClearProgress(<%= context.namespaceAppCamelCase %>, dtoIn, config, lockSecret, session) {
    const uuTerritoryUri = this._parseTerritoryUri(<%= context.namespaceAppCamelCase %>.uuTerritoryBaseUri);

    let progressClient;
    let progress = {
      expireAt: UuDateTime.now().shiftDay(7),
      name: <%= context.projectClassName %>Constants.getClearProgressName(<%= context.namespaceAppCamelCase %>.awid),
      code: <%= context.projectClassName %>Constants.getClearProgressCode(<%= context.namespaceAppCamelCase %>.awid),
      authorizationStrategy: "boundArtifact",
      boundArtifactUri: uuTerritoryUri.setParameter("id", <%= context.namespaceAppCamelCase %>.artifactId).toUri().toString(),
      boundArtifactPermissionMatrix: <%= context.projectClassName %>Constants.CONSOLE_BOUND_MATRIX,
      lockSecret,
    };

    try {
      progressClient = await ProgressClient.get(config.uuConsoleBaseUri, { code: progress.code }, { session });
    } catch (e) {
      if (e.cause?.code !== ProgressConstants.PROGRESS_DOES_NOT_EXIST) {
        throw new Errors.Clear.ProgressGetCallFailed({ progressCode: progress.code }, e);
      }
    }

    if (!progressClient) {
      try {
        progressClient = await ProgressClient.createInstance(config.uuConsoleBaseUri, progress, { session });
      } catch (e) {
        throw new Errors.Clear.ProgressCreateCallFailed({ progressCode: progress.code }, e);
      }
    } else if (dtoIn.force) {
      try {
        await progressClient.releaseLock();
      } catch (e) {
        if (e.cause?.code !== ProgressConstants.PROGRESS_RELEASE_DOES_NOT_EXIST) {
          throw new Errors.Clear.ProgressReleaseLockCallFailed({ progressCode: progress.code }, e);
        }
      }

      try {
        await progressClient.setState({ state: "cancelled" });
      } catch (e) {
        DtoBuilder.addWarning(new Warnings.Clear.ProgressSetStateCallFailed(e.cause?.paramMap));
      }

      try {
        await progressClient.delete();
      } catch (e) {
        if (e.cause?.code !== ProgressConstants.PROGRESS_DELETE_DOES_NOT_EXIST) {
          throw new Errors.Clear.ProgressDeleteCallFailed({ progressCode: progress.code }, e);
        }
      }

      try {
        progressClient = await ProgressClient.createInstance(config.uuConsoleBaseUri, progress, { session });
      } catch (e) {
        throw new Errors.Clear.ProgressCreateCallFailed({ progressCode: progress.code }, e);
      }
    }

    try {
      await progressClient.start({
        message: "Progress was started",
        totalWork: <%= context.projectClassName %>Constants.getClearStepCount(),
        lockSecret,
      });
    } catch (e) {
      throw new Errors.Clear.ProgressStartCallFailed({ progressCode: progress.code }, e);
    }

    return progressClient;
  }

  async _setClearProgressAuthorizationStrategy(uuConsoleBaseUri, progressCode, awid, awidInitiatorList, session) {
    let progressClient;

    try {
      progressClient = await ProgressClient.get(uuConsoleBaseUri, { code: progressCode }, { session });
    } catch (e) {
      if (e.cause?.code === ProgressConstants.PROGRESS_DOES_NOT_EXIST) {
        return;
      } else {
        throw new Errors._clearFinalize.ProgressGetCallFailed({ code: progressCode }, e);
      }
    }

    try {
      await progressClient.setAuthorizationStrategy({
        authorizationStrategy: "uuIdentityList",
        permissionMap: await this._getClearProgressPermissionMap(awid, awidInitiatorList, session),
        force: true,
      });
    } catch (e) {
      if (e.cause?.code !== ProgressConstants.PROGRESS_HAS_SAME_AUTH_STRATEGY) {
        throw new Errors._clearFinalize.ProgressSetAuthorizationStrategyCallFailed({ code: progressCode }, e);
      }
    }
  }

  async _getClearProgressPermissionMap(awid, awidInitiatorList, sysIdentitySession) {
    const awidData = await UuAppWorkspace.get(awid);

    let permissionMap = {};
    for (let identity of Array.from(new Set([...awidData.awidInitiatorList, ...awidInitiatorList]))) {
      permissionMap[identity] = <%= context.projectClassName %>Constants.CONSOLE_BOUND_MATRIX.Authorities;
    }
    permissionMap[sysIdentitySession.getIdentity().getUuIdentity()] =
      <%= context.projectClassName %>Constants.CONSOLE_BOUND_MATRIX.Authorities;

    return permissionMap;
  }

  async _deleteProgress(progressCode, uuConsoleBaseUri, session) {
    let progressClient;

    try {
      progressClient = await ProgressClient.get(uuConsoleBaseUri, { code: progressCode }, { session });
    } catch (e) {
      if (e.cause?.code === ProgressConstants.PROGRESS_DOES_NOT_EXIST) {
        return;
      } else {
        throw new Errors.Clear.ProgressGetCallFailed({ code: progressCode }, e);
      }
    }

    try {
      await progressClient.setState({ state: "final" });
      await progressClient.delete();
    } catch (e) {
      DtoBuilder.addWarning(new Warnings._clearFinalize.FailedToDeleteProgress(e.parameters));
    }
  }

  async _clearConsole(uuConsoleBaseUri, consoleCode, session) {
    const consoleClient = new ConsoleClient(uuConsoleBaseUri, { code: consoleCode }, { session });

    try {
      await consoleClient.clear();
    } catch (e) {
      DtoBuilder.addWarning(new Warnings._clearFinalize.FailedToClearConsole({ code: consoleCode }));
    }
  }

  async _runScript(appUri, dtoIn, configuration, lockSecret, session) {
    const scriptEngineClient = new ScriptEngineClient({
      scriptEngineUri: configuration.uuScriptEngineBaseUri,
      consoleUri: configuration.uuConsoleBaseUri,
      consoleCode: <%= context.projectClassName %>Constants.getMainConsoleCode(appUri.getAwid()),
      scriptRepositoryUri: configuration.uuScriptRepositoryBaseUri,
      session,
    });

    const scriptDtoIn = {
      <%= context.namespaceAppCamelCase %>Uri: appUri.toString(),
      awidInitiatorList: dtoIn.data.awidInitiatorList,
      lockSecret,
    };

    await scriptEngineClient.runScript({ scriptCode: SCRIPT_CODE, scriptDtoIn });
  }
}

module.exports = new ClearAbl();
