// uu_appg01_devkit-common/src/templates/__nodejs-app/app/abl/project/init-abl.js
"use strict";
const Crypto = require("crypto");
const { DaoFactory } = require("uu_appg01_server").ObjectStore;
const { UuAppWorkspace } = require("uu_appg01_server").Workspace;
const { AuthenticationService } = require("uu_appg01_server").Authentication;
const { UriBuilder } = require("uu_appg01_server").Uri;
const { UuDateTime } = require("uu_i18ng01");
const { ConsoleClient, ProgressClient } = require("uu_consoleg02-uulib");

const Errors = require("../../api/errors/<%= context.projectClassFile %>-error");
const Warnings = require("../../api/warnings/<%= context.projectClassFile %>-warning");
const Validator = require("../../components/validator");
const DtoBuilder = require("../../components/dto-builder");
const ScriptEngineClient = require("../../components/script-engine-client");
const <%= context.projectClassName %>Client = require("../../components/<%= context.projectClassFile %>-client");
const StepHandler = require("../../components/step-handler");
const InitRollbackAbl = require("./init-rollback-abl");

const ConsoleConstants = require("../../constants/console-constants");
const ProgressConstants = require("../../constants/progress-constants");
const <%= context.projectClassName %>Constants = require("../../constants/<%= context.projectClassFile %>-constants");
const Configuration = require("../../components/configuration");

const SCRIPT_CODE = "<%=  context.nameWithVendor.replace(/-server$/, "-uuscriptlib") %>/<%= context.projectClassFile%>/init";

class InitAbl {
  constructor() {
    this.dao = DaoFactory.getDao(<%= context.projectClassName %>Constants.Schemas.<%=  context.app.toUpperCase() %>_INSTANCE);
  }

  async init(uri, dtoIn) {
    // HDS 1
    const awid = uri.getAwid();
    this._validateDtoIn(uri, dtoIn);

    // HDS 2
    let <%= context.namespaceAppCamelCase %> = await this.dao.getByAwid(awid);
    let uuAppWorkspace = await UuAppWorkspace.get(awid);

    // HDS 3
    this._validateMode(<%= context.namespaceAppCamelCase %>, dtoIn, uuAppWorkspace.sysState);

    // HDS 4
    const configuration = await Configuration.getUuSubAppConfiguration({
      awid,
      artifactId: dtoIn.data.locationId || <%= context.namespaceAppCamelCase %>.temporaryData.dtoIn.locationId,
      uuTerritoryBaseUri: dtoIn.data.uuTerritoryBaseUri || <%= context.namespaceAppCamelCase %>.temporaryData.dtoIn.uuTerritoryBaseUri,
    });

    // HDS 5
    let initData;
    switch (dtoIn.mode) {
      case <%= context.projectClassName %>Constants.ModeMap.STANDARD: {
        initData = dtoIn.data;
        const uuTerritoryBaseUri = this._parseTerritoryUri(initData.uuTerritoryBaseUri);
        const temporaryData = {
          useCase: uri.getUseCase(),
          dtoIn: { ...initData },
          stepList: [<%= context.projectClassName %>Constants.InitStepMap.<%=  context.app.toUpperCase() %>_OBJECT_CREATED.code],
          progressMap: {
            uuConsoleUri: configuration.uuConsoleBaseUri,
            progressCode: <%= context.projectClassName %>Constants.getInitProgressCode(awid),
            consoleCode: <%= context.projectClassName %>Constants.getMainConsoleCode(awid),
          },
        };

        <%= context.namespaceAppCamelCase %> = await this.dao.create({
          awid,
          state: <%= context.projectClassName %>Constants.StateMap.CREATED,
          code: `${<%=  context.projectClassName %>Constants.AWSC_PREFIX}/${awid}`,
          uuTerritoryBaseUri: uuTerritoryBaseUri.toString(),
          name: initData.name,
          desc: initData.desc,
          temporaryData,
        });

        try {
          await UuAppWorkspace.setBeingInitializedSysState(awid);
        } catch (e) {
          throw new Errors.Init.SetBeingInitializedSysStateFailed({}, e);
        }
        break;
      }

      case <%= context.projectClassName %>Constants.ModeMap.RETRY: {
        initData = <%= context.namespaceAppCamelCase %>.temporaryData.dtoIn;
        break;
      }

      case <%= context.projectClassName %>Constants.ModeMap.ROLLBACK: {
        <%= context.namespaceAppCamelCase %>.temporaryData.rollbackMode = true;
        if (!<%= context.namespaceAppCamelCase %>.temporaryData.rollbackStepList) {
          <%= context.namespaceAppCamelCase %>.temporaryData.rollbackStepList = [];
        }
        <%= context.namespaceAppCamelCase %> = await this.dao.updateByAwid({ ...<%= context.namespaceAppCamelCase %> });
        initData = <%= context.namespaceAppCamelCase %>.temporaryData.dtoIn;
        break;
      }

      default: {
        throw new Errors.Init.WrongModeAndCircumstances({
          mode: dtoIn.mode,
          appObjectState: <%= context.namespaceAppCamelCase %>?.state,
          temporaryData: <%= context.namespaceAppCamelCase %>?.temporaryData,
        });
      }
    }

    // HDS 6
    const sysIdentitySession = await AuthenticationService.authenticateSystemIdentity();
    const lockSecret = Crypto.randomBytes(32).toString("hex");
    const progressClient = await this._createInitProgress(
      <%= context.namespaceAppCamelCase %>,
      dtoIn,
      configuration,
      lockSecret,
      sysIdentitySession,
    );

    // HDS 7
    switch (dtoIn.mode) {
      case <%= context.projectClassName %>Constants.ModeMap.STANDARD:
      case <%= context.projectClassName %>Constants.ModeMap.RETRY: {
        const stepHandler = new StepHandler({
          schema: <%= context.projectClassName %>Constants.Schemas.<%=  context.app.toUpperCase() %>_INSTANCE,
          progressClient,
          stepList: <%= context.namespaceAppCamelCase %>?.temporaryData?.stepList,
        });

        const <%= context.projectClassName.charAt(0).toLowerCase() + context.projectClassName.substr(1) %>Client = new <%= context.projectClassName %>Client(<%= context.namespaceAppCamelCase %>, <%= context.namespaceAppCamelCase %>.uuTerritoryBaseUri);

        <%= context.namespaceAppCamelCase %> = await stepHandler.handleStep(<%= context.namespaceAppCamelCase %>, <%= context.projectClassName %>Constants.InitStepMap.AWSC_CREATED, async () => {
          <%= context.namespaceAppCamelCase %>.state = <%= context.projectClassName %>Constants.StateMap.BEING_INITIALIZED;
          await this.dao.updateByAwid({ ...<%= context.namespaceAppCamelCase %> });
          await <%= context.projectClassName.charAt(0).toLowerCase() + context.projectClassName.substr(1) %>Client.createAwsc(
            initData.locationId,
            initData.responsibleRoleId,
            initData.permissionMatrix,
            configuration.uuAppMetaModelVersion,
          );
        });

        <%= context.namespaceAppCamelCase %> = await stepHandler.handleStep(<%= context.namespaceAppCamelCase %>, <%= context.projectClassName %>Constants.InitStepMap.WS_CONNECTED, async () => {
          await this._connectAwsc(<%= context.namespaceAppCamelCase %>, uri.getBaseUri(), <%= context.namespaceAppCamelCase %>.uuTerritoryBaseUri, sysIdentitySession);
        });

        <%= context.namespaceAppCamelCase %> = await stepHandler.handleStep(<%= context.namespaceAppCamelCase %>, <%= context.projectClassName %>Constants.InitStepMap.CONSOLE_CREATED, async () => {
          await this._createConsole(<%= context.namespaceAppCamelCase %>, configuration, sysIdentitySession);
        });

        // TODO If your application requires any additional steps, add them here...

        if (!<%= context.namespaceAppCamelCase %>.temporaryData.stepList.includes(<%= context.projectClassName %>Constants.InitStepMap.PROGRESS_ENDED.code)) {
          await this._runScript(uri.getBaseUri(), configuration, lockSecret, sysIdentitySession);
        } else {
          await this._initFinalize(uri, { lockSecret });
        }
        break;
      }

      case <%= context.projectClassName %>Constants.ModeMap.ROLLBACK: {
        if (
          <%= context.namespaceAppCamelCase %>.temporaryData.stepList.includes(<%= context.projectClassName %>Constants.InitStepMap.CONSOLE_CREATED.code) &&
          !<%= context.namespaceAppCamelCase %>.temporaryData.rollbackStepList.includes(<%= context.projectClassName %>Constants.InitRollbackStepMap.CONSOLE_CLEARED.code)
        ) {
          await InitRollbackAbl.initRollback(uri.getBaseUri(), configuration, lockSecret);
        } else {
          await InitRollbackAbl._initFinalizeRollback(uri, { lockSecret });
        }
        break;
      }

      default: {
        throw new Errors.Init.WrongModeAndCircumstances({
          mode: dtoIn.mode,
          appObjectState: <%= context.namespaceAppCamelCase %>?.state,
          temporaryData: <%= context.namespaceAppCamelCase %>?.temporaryData,
        });
      }
    }

    // HDS 8
    return DtoBuilder.prepareDtoOut({ data: <%= context.namespaceAppCamelCase %> });
  }

  async _initFinalize(uri, dtoIn) {
    // HDS 1
    const awid = uri.getAwid();
    Validator.validateDtoInCustom(uri, dtoIn, "sysUuAppWorkspaceInitFinalizeDtoInType");

    // HDS 2
    let <%= context.namespaceAppCamelCase %> = await this.dao.getByAwid(awid);

    if (!<%= context.namespaceAppCamelCase %>) {
      // 2.1
      throw new Errors._initFinalize.<%= context.namespaceAppCamelCase.charAt(0).toUpperCase() + context.namespaceAppCamelCase.slice(1) %>DoesNotExist({ awid });
    }

    if (![<%= context.projectClassName %>Constants.StateMap.BEING_INITIALIZED, <%= context.projectClassName %>Constants.StateMap.ACTIVE].includes(<%= context.namespaceAppCamelCase %>.state)) {
      // 2.2
      throw new Errors._initFinalize.NotInProperState({
        state: <%= context.namespaceAppCamelCase %>.state,
        expectedStateList: [<%= context.projectClassName %>Constants.StateMap.BEING_INITIALIZED, <%= context.projectClassName %>Constants.StateMap.ACTIVE],
      });
    }

    // HDS 3
    const sysIdentitySession = await AuthenticationService.authenticateSystemIdentity();
    const progress = {
      code: <%= context.projectClassName %>Constants.getInitProgressCode(<%= context.namespaceAppCamelCase %>.awid),
      lockSecret: dtoIn.lockSecret,
    };
    let progressClient = null;
    if (!<%= context.namespaceAppCamelCase %>.temporaryData.stepList.includes(<%= context.projectClassName %>Constants.InitStepMap.PROGRESS_ENDED.code)) {
      progressClient = await ProgressClient.get(<%= context.namespaceAppCamelCase %>.temporaryData.progressMap.uuConsoleUri, progress, {
        session: sysIdentitySession,
      });
    }
    const stepHandler = new StepHandler({
      schema: <%= context.projectClassName %>Constants.Schemas.<%=  context.app.toUpperCase() %>_INSTANCE,
      progressClient,
      stepList: <%= context.namespaceAppCamelCase %>.temporaryData.stepList,
    });

    // TODO If your application requires any additional steps, add them here...

    // HDS 4
    <%= context.namespaceAppCamelCase %> = await stepHandler.handleStep(
      <%= context.namespaceAppCamelCase %>,
      <%= context.projectClassName %>Constants.InitStepMap.PROGRESS_ENDED,
      async () => {
        await progressClient.end({
          state: ProgressConstants.StateMap.COMPLETED,
          message: "Initialization finished.",
          doneWork: <%= context.projectClassName %>Constants.getInitStepCount(),
        });
      },
      false,
    );

    // HDS 5
    if (<%= context.namespaceAppCamelCase %>.state === <%= context.projectClassName %>Constants.StateMap.BEING_INITIALIZED) {
      <%= context.namespaceAppCamelCase %> = await this.dao.updateByAwid({ awid, state: <%= context.projectClassName %>Constants.StateMap.ACTIVE, temporaryData: null });
    }

    // HDS 6
    await UuAppWorkspace.setActiveSysState(awid);

    // HDS 7
    return DtoBuilder.prepareDtoOut({ data: <%= context.namespaceAppCamelCase %> });
  }

  // Validates dtoIn. In case of standard mode the data key of dtoIn is also validated.
  _validateDtoIn(uri, dtoIn) {
    let uuAppErrorMap = Validator.validateDtoIn(uri, dtoIn);
    if (dtoIn.mode === <%= context.projectClassName %>Constants.ModeMap.STANDARD) {
      Validator.validateDtoInCustom(uri, dtoIn.data, "sysUuAppWorkspaceInitStandardDtoInType", uuAppErrorMap);
    }
    return uuAppErrorMap;
  }

  _validateMode(<%= context.namespaceAppCamelCase %>, dtoIn, sysState) {
    switch (dtoIn.mode) {
      case <%= context.projectClassName %>Constants.ModeMap.STANDARD:
        if (![UuAppWorkspace.SYS_STATES.ASSIGNED, UuAppWorkspace.SYS_STATES.BEING_INITIALIZED].includes(sysState)) {
          // 3.A.1.1.
          throw new Errors.Init.SysUuAppWorkspaceIsNotInProperState({
            sysState,
            expectedSysStateList: [UuAppWorkspace.SYS_STATES.ASSIGNED, UuAppWorkspace.SYS_STATES.BEING_INITIALIZED],
          });
        }
        if (<%= context.namespaceAppCamelCase %>) {
          // 3.A.2.1.
          throw new Errors.Init.<%= context.namespaceAppCamelCase.charAt(0).toUpperCase() + context.namespaceAppCamelCase.slice(1) %>ObjectAlreadyExist({
            mode: dtoIn.mode,
            allowedModeList: [<%= context.projectClassName %>Constants.ModeMap.RETRY, <%= context.projectClassName %>Constants.ModeMap.ROLLBACK],
          });
        }
        break;

      case <%= context.projectClassName %>Constants.ModeMap.RETRY:
        if (sysState !== UuAppWorkspace.SYS_STATES.BEING_INITIALIZED) {
          // 3.B.1.1.
          throw new Errors.Init.SysUuAppWorkspaceIsNotInProperState({
            sysState,
            expectedSysStateList: [UuAppWorkspace.SYS_STATES.BEING_INITIALIZED],
          });
        }
        if (!<%= context.namespaceAppCamelCase %>?.temporaryData) {
          // 3.B.2.1.
          throw new Errors.Init.MissingRequiredData();
        }
        if (<%= context.namespaceAppCamelCase %>?.temporaryData?.rollbackMode) {
          // 3.B.3.1.
          throw new Errors.Init.RollbackNotFinished();
        }
        break;

      case <%= context.projectClassName %>Constants.ModeMap.ROLLBACK:
        if (sysState !== UuAppWorkspace.SYS_STATES.BEING_INITIALIZED) {
          // 3.C.1.1.
          throw new Errors.Init.SysUuAppWorkspaceIsNotInProperState({
            sysState,
            expectedSysStateList: [UuAppWorkspace.SYS_STATES.BEING_INITIALIZED],
          });
        }
        if (!<%= context.namespaceAppCamelCase %>?.temporaryData) {
          // 3.C.2.1.
          throw new Errors.Init.MissingRequiredData();
        }
        if (!dtoIn.force && <%= context.namespaceAppCamelCase %>?.temporaryData?.rollbackMode) {
          // 3.C.3.1.
          throw new Errors.Init.RollbackAlreadyRunning();
        }
        break;
    }
  }

  _parseTerritoryUri(locationUri) {
    let uuTerritoryUri;

    try {
      uuTerritoryUri = UriBuilder.parse(locationUri).toUri();
    } catch (e) {
      throw new Errors.Init.UuTLocationUriParseFailed({ uri: locationUri }, e);
    }

    return uuTerritoryUri.getBaseUri();
  }

  async _createInitProgress(<%= context.namespaceAppCamelCase %>, dtoIn, config, lockSecret, session) {
    let progressClient;
    let progress = {
      expireAt: UuDateTime.now().shiftDay(7),
      name: <%= context.projectClassName %>Constants.getInitProgressName(<%= context.namespaceAppCamelCase %>.awid),
      code: <%= context.projectClassName %>Constants.getInitProgressCode(<%= context.namespaceAppCamelCase %>.awid),
      authorizationStrategy: "uuIdentityList",
      permissionMap: await this._getInitProgressPermissionMap(<%= context.namespaceAppCamelCase %>.awid, session),
      lockSecret,
    };

    try {
      progressClient = await ProgressClient.get(config.uuConsoleBaseUri, { code: progress.code }, { session });
    } catch (e) {
      if (e.cause?.code !== ProgressConstants.PROGRESS_DOES_NOT_EXIST) {
        throw new Errors.Init.ProgressGetCallFailed({ progressCode: progress.code }, e);
      }
    }

    if (!progressClient) {
      try {
        progressClient = await ProgressClient.createInstance(config.uuConsoleBaseUri, progress, { session });
      } catch (e) {
        throw new Errors.Init.ProgressCreateCallFailed({ progressCode: progress.code }, e);
      }
    } else if (dtoIn.force) {
      try {
        await progressClient.releaseLock();
      } catch (e) {
        if (e.cause?.code !== ProgressConstants.PROGRESS_RELEASE_DOES_NOT_EXIST) {
          throw new Errors.Init.ProgressReleaseLockCallFailed({ progressCode: progress.code }, e);
        }
      }

      try {
        await progressClient.setState({ state: "cancelled" });
      } catch (e) {
        DtoBuilder.addWarning(new Warnings.Init.ProgressSetStateCallFailed(e.cause?.paramMap));
      }

      try {
        await progressClient.delete();
      } catch (e) {
        if (e.cause?.code !== ProgressConstants.PROGRESS_DELETE_DOES_NOT_EXIST) {
          throw new Errors.Init.ProgressDeleteCallFailed({ progressCode: progress.code }, e);
        }
      }

      try {
        progressClient = await ProgressClient.createInstance(config.uuConsoleBaseUri, progress, { session });
      } catch (e) {
        throw new Errors.Init.ProgressCreateCallFailed({ progressCode: progress.code }, e);
      }
    }

    try {
      await progressClient.start({
        message: "Progress was started",
        totalWork:
          dtoIn.mode === <%= context.projectClassName %>Constants.ModeMap.ROLLBACK
            ? <%= context.projectClassName %>Constants.getInitRollbackStepCount()
            : <%= context.projectClassName %>Constants.getInitStepCount(),
        lockSecret,
      });
    } catch (e) {
      throw new Errors.Init.ProgressStartCallFailed({ progressCode: progress.code }, e);
    }

    return progressClient;
  }

  async _getInitProgressPermissionMap(awid, sysIdentitySession) {
    const awidData = await UuAppWorkspace.get(awid);

    let permissionMap = {};
    for (let identity of awidData.awidInitiatorList) {
      permissionMap[identity] = <%= context.projectClassName %>Constants.CONSOLE_BOUND_MATRIX.Authorities;
    }
    permissionMap[sysIdentitySession.getIdentity().getUuIdentity()] =
      <%= context.projectClassName %>Constants.CONSOLE_BOUND_MATRIX.Authorities;

    return permissionMap;
  }

  async _connectAwsc(<%= context.namespaceAppCamelCase %>, appUri, uuTerritoryBaseUri, session) {
    const artifactUri = UriBuilder.parse(uuTerritoryBaseUri).setParameter("id", <%= context.namespaceAppCamelCase %>.artifactId).toUri().toString();

    try {
      await UuAppWorkspace.connectArtifact(appUri, { artifactUri }, session);
    } catch (e) {
      throw new Errors.<%= context.projectClassName %>.ConnectAwscFailed(
        {
          awid: <%= context.namespaceAppCamelCase %>.awid,
          awscId: <%= context.namespaceAppCamelCase %>.artifactId,
          uuTerritoryBaseUri,
        },
        e,
      );
    }
  }

  async _createConsole(<%= context.namespaceAppCamelCase %>, configuration, session) {
    const artifactUri = UriBuilder.parse(<%= context.namespaceAppCamelCase %>.uuTerritoryBaseUri).setParameter("id", <%= context.namespaceAppCamelCase %>.artifactId).toString();
    const console = {
      code: <%= context.projectClassName %>Constants.getMainConsoleCode(<%= context.namespaceAppCamelCase %>.awid),
      authorizationStrategy: "boundArtifact",
      boundArtifactUri: artifactUri,
      boundArtifactPermissionMatrix: <%= context.projectClassName %>Constants.CONSOLE_BOUND_MATRIX,
    };

    try {
      await ConsoleClient.createInstance(configuration.uuConsoleBaseUri, console, { session });
    } catch (e) {
      throw new Errors.Init.FailedToCreateConsole({}, e);
    }
  }

  async _setConsoleExpiration(uuConsoleUri, consoleCode, session) {
    let consoleClient;
    try {
      consoleClient = await ConsoleClient.get(uuConsoleUri, { code: consoleCode }, { session });
    } catch (e) {
      if (e.cause?.code === ConsoleConstants.CONSOLE_GET_DOES_NOT_EXISTS) {
        throw new Errors._initFinalize.ConsoleGetCallFailed({ code: consoleCode }, e);
      }
    }

    try {
      await consoleClient.update({ expireAt: new UuDateTime().shiftDay(7).date });
    } catch (e) {
      if (e.cause?.code === ConsoleConstants.CONSOLE_UPDATE_DOES_NOT_EXISTS) {
        DtoBuilder.addWarning(new Warnings._initFinalize.ConsoleDoesNotExist({ code: consoleCode }));
      } else {
        throw new Errors._initFinalize.ConsoleUpdateCallFailed({ code: consoleCode }, e);
      }
    }
  }

  async _runScript(appUri, configuration, lockSecret, session) {
    const scriptEngineClient = new ScriptEngineClient({
      scriptEngineUri: configuration.uuScriptEngineBaseUri,
      consoleUri: configuration.uuConsoleBaseUri,
      consoleCode: <%= context.projectClassName %>Constants.getMainConsoleCode(appUri.getAwid()),
      scriptRepositoryUri: configuration.uuScriptRepositoryBaseUri,
      session,
    });

    const scriptDtoIn = {
      <%= context.namespaceAppCamelCase %>Uri: appUri.toString(),
      lockSecret,
    };

    await scriptEngineClient.runScript({ scriptCode: SCRIPT_CODE, scriptDtoIn });
  }
}

module.exports = new InitAbl();
