// uu_appg01_devkit-common/src/templates/__nodejs-app/app/abl/project/set-state-closed-abl.js
"use strict";
const Crypto = require("crypto");
const { DaoFactory } = require("uu_appg01_server").ObjectStore;
const { AuthenticationService } = require("uu_appg01_server").Authentication;
const { UriBuilder } = require("uu_appg01_server").Uri;
const { UuDateTime } = require("uu_i18ng01");
const { ProgressClient } = require("uu_consoleg02-uulib");

const Errors = require("../../api/errors/<%= context.projectClassFile %>-error");
const Warnings = require("../../api/warnings/<%= context.projectClassFile %>-warning");
const Validator = require("../../components/validator");
const DtoBuilder = require("../../components/dto-builder");
const ScriptEngineClient = require("../../components/script-engine-client");
const <%= context.projectClassName %>Client = require("../../components/<%= context.projectClassFile %>-client");
const StepHandler = require("../../components/step-handler");

const ProgressConstants = require("../../constants/progress-constants");
const <%= context.projectClassName %>Constants = require("../../constants/<%= context.projectClassFile %>-constants");
const TerritoryConstants = require("../../constants/territory-constants");
const Configuration = require("../../components/configuration");

const SCRIPT_CODE = "<%=  context.nameWithVendor.replace(/-server$/, "-uuscriptlib") %>/<%= context.projectClassFile%>/set-state-closed";

class SetStateClosedAbl {
  constructor() {
    this.dao = DaoFactory.getDao(<%= context.projectClassName %>Constants.Schemas.<%=  context.app.toUpperCase() %>_INSTANCE);
  }

  async setStateClosed(uri, dtoIn) {
    // HDS 1
    const awid = uri.getAwid();
    Validator.validateDtoIn(uri, dtoIn);

    // HDS 2
    let <%= context.namespaceAppCamelCase %> = await this.dao.getByAwid(awid);

    if (!<%= context.namespaceAppCamelCase %>) {
      // 2.1
      throw new Errors.SetStateClosed.<%= context.namespaceAppCamelCase.charAt(0).toUpperCase() + context.namespaceAppCamelCase.slice(1) %>DoesNotExist({ awid });
    }

    if (<%= context.namespaceAppCamelCase %>.state !== <%= context.projectClassName %>Constants.StateMap.ACTIVE) {
      // 2.2
      throw new Errors.SetStateClosed.NotInProperState({
        state: <%= context.namespaceAppCamelCase %>.state,
        expectedStateList: [<%= context.projectClassName %>Constants.StateMap.ACTIVE],
      });
    }

    if (<%= context.namespaceAppCamelCase %>.temporaryData && <%= context.namespaceAppCamelCase %>.temporaryData.useCase !== uri.getUseCase()) {
      // 2.3
      throw new Errors.SetStateClosed.UseCaseExecutionForbidden({
        concurrencyUseCase: <%= context.namespaceAppCamelCase %>.temporaryData.useCase,
      });
    }

    // HDS 3
    const configuration = await Configuration.getUuSubAppConfiguration({
      awid,
      artifactId: <%= context.namespaceAppCamelCase %>.artifactId,
      uuTerritoryBaseUri: <%= context.namespaceAppCamelCase %>.uuTerritoryBaseUri,
    });

    // HDS 4
    const sysIdentitySession = await AuthenticationService.authenticateSystemIdentity();
    const lockSecret = Crypto.randomBytes(32).toString("hex");
    const progressClient = await this._createSetStateClosedProgress(
      <%= context.namespaceAppCamelCase %>,
      dtoIn,
      configuration,
      lockSecret,
      sysIdentitySession,
    );

    // HDS 5
    if (!<%= context.namespaceAppCamelCase %>.temporaryData) {
      <%= context.namespaceAppCamelCase %> = await this.dao.updateByAwid({
        awid,
        temporaryData: {
          useCase: uri.getUseCase(),
          dtoIn: {},
          stepList: [<%= context.projectClassName %>Constants.SetStateClosedStepMap.CLOSE_STARTED.code],
          progressMap: {
            progressCode: progressClient.progress.code,
            uuConsoleUri: configuration.uuConsoleBaseUri,
            consoleCode: <%= context.projectClassName %>Constants.getMainConsoleCode(awid),
          },
        },
      });
    }

    // TODO If your application requires any additional steps, add them here...

    // HDS 6
    await this._runScript(uri.getBaseUri(), configuration, progressClient.progress.lockSecret, sysIdentitySession);

    // HDS 7
    return DtoBuilder.prepareDtoOut({ data: <%= context.namespaceAppCamelCase %> });
  }

  async _setStateClosedFinalize(uri, dtoIn) {
    // HDS 1
    const awid = uri.getAwid();
    Validator.validateDtoInCustom(uri, dtoIn, "sysUuAppWorkspaceSetStateClosedFinalizeDtoInType");

    // HDS 2
    let <%= context.namespaceAppCamelCase %> = await this.dao.getByAwid(awid);

    if (!<%= context.namespaceAppCamelCase %>) {
      // 2.1
      throw new Errors._setStateClosedFinalize.<%= context.namespaceAppCamelCase.charAt(0).toUpperCase() + context.namespaceAppCamelCase.slice(1) %>DoesNotExist({ awid });
    }

    if (!<%= context.namespaceAppCamelCase %>.state === <%= context.projectClassName %>Constants.StateMap.ACTIVE) {
      // 2.2
      throw new Errors._setStateClosedFinalize.NotInProperState({
        state: <%= context.namespaceAppCamelCase %>.state,
        expectedStateList: [<%= context.projectClassName %>Constants.StateMap.ACTIVE],
      });
    }

    if (!<%= context.namespaceAppCamelCase %>.temporaryData) {
      // 2.3
      throw new Errors._setStateClosedFinalize.MissingRequiredData();
    }

    if (<%= context.namespaceAppCamelCase %>.temporaryData && <%= context.namespaceAppCamelCase %>.temporaryData.useCase !== "sys/uuAppWorkspace/setStateClosed") {
      // 2.4
      throw new Errors._setStateClosedFinalize.UseCaseExecutionForbidden({
        concurrencyUseCase: <%= context.namespaceAppCamelCase %>.temporaryData.useCase,
      });
    }

    // HDS 3
    const sysIdentitySession = await AuthenticationService.authenticateSystemIdentity();
    const progress = {
      code: <%= context.projectClassName %>Constants.getSetStateClosedProgressCode(<%= context.namespaceAppCamelCase %>.awid),
      lockSecret: dtoIn.lockSecret,
    };
    let progressClient = null;
    if (!<%= context.namespaceAppCamelCase %>.temporaryData.stepList.includes(<%= context.projectClassName %>Constants.SetStateClosedStepMap.PROGRESS_ENDED.code)) {
      progressClient = await ProgressClient.get(<%= context.namespaceAppCamelCase %>.temporaryData.progressMap.uuConsoleUri, progress, {
        session: sysIdentitySession,
      });
    }
    const stepHandler = new StepHandler({
      schema: <%= context.projectClassName %>Constants.Schemas.<%=  context.app.toUpperCase() %>_INSTANCE,
      progressClient,
      stepList: <%= context.namespaceAppCamelCase %>.temporaryData.stepList,
    });

    // TODO If your application requires any additional steps, add them here...

    // HDS 4
    <%= context.namespaceAppCamelCase %> = await stepHandler.handleStep(<%= context.namespaceAppCamelCase %>, <%= context.projectClassName %>Constants.SetStateClosedStepMap.AWSC_CLOSED, async () => {
      const <%= context.projectClassName.charAt(0).toLowerCase() + context.projectClassName.substr(1) %>Client = new <%= context.projectClassName %>Client(<%= context.namespaceAppCamelCase %>, <%= context.namespaceAppCamelCase %>.uuTerritoryBaseUri);
      try {
        await <%= context.projectClassName.charAt(0).toLowerCase() + context.projectClassName.substr(1) %>Client.setAwscState(<%= context.projectClassName %>Constants.StateMap.FINAL);
      } catch (e) {
        if (e.cause?.code !== TerritoryConstants.INVALID_ARTIFACT_STATE) {
          throw e;
        } else {
          DtoBuilder.addWarning(new Warnings._setStateClosedFinalize.AwscAlreadyInFinalState());
        }
      }
    });

    // HDS 5
    <%= context.namespaceAppCamelCase %> = await stepHandler.handleStep(
      <%= context.namespaceAppCamelCase %>,
      <%= context.projectClassName %>Constants.SetStateClosedStepMap.PROGRESS_ENDED,
      async () => {
        await progressClient.end({
          state: ProgressConstants.StateMap.COMPLETED,
          message: "Setting closed state finished.",
          expireAt: UuDateTime.now().shiftDay(7),
          doneWork: <%= context.projectClassName %>Constants.getSetStateClosedStepCount(),
        });
      },
      false,
    );

    // HDS 6
    <%= context.namespaceAppCamelCase %> = await this.dao.updateByAwid({
      awid,
      state: <%= context.projectClassName %>Constants.StateMap.FINAL,
      temporaryData: null,
    });

    // HDS 7
    return DtoBuilder.prepareDtoOut();
  }

  _parseTerritoryUri(locationUri) {
    let uuTerritoryUri;

    try {
      uuTerritoryUri = UriBuilder.parse(locationUri);
    } catch (e) {
      throw new Errors.SetStateClosed.UuTLocationUriParseFailed({ uri: locationUri }, e);
    }

    return uuTerritoryUri;
  }

  async _createSetStateClosedProgress(<%= context.namespaceAppCamelCase %>, dtoIn, config, lockSecret, session) {
    const uuTerritoryUri = this._parseTerritoryUri(<%= context.namespaceAppCamelCase %>.uuTerritoryBaseUri);

    let progressClient;
    let progress = {
      expireAt: UuDateTime.now().shiftDay(7),
      name: <%= context.projectClassName %>Constants.getSetStateClosedProgressName(<%= context.namespaceAppCamelCase %>.awid),
      code: <%= context.projectClassName %>Constants.getSetStateClosedProgressCode(<%= context.namespaceAppCamelCase %>.awid),
      authorizationStrategy: "boundArtifact",
      boundArtifactUri: uuTerritoryUri.setParameter("id", <%= context.namespaceAppCamelCase %>.artifactId).toUri().toString(),
      boundArtifactPermissionMatrix: <%= context.projectClassName %>Constants.CONSOLE_BOUND_MATRIX,
      lockSecret,
    };

    try {
      progressClient = await ProgressClient.get(config.uuConsoleBaseUri, { code: progress.code }, { session });
    } catch (e) {
      if (e.cause?.code !== ProgressConstants.PROGRESS_DOES_NOT_EXIST) {
        throw new Errors.SetStateClosed.ProgressGetCallFailed({ progressCode: progress.code }, e);
      }
    }

    if (!progressClient) {
      try {
        progressClient = await ProgressClient.createInstance(config.uuConsoleBaseUri, progress, { session });
      } catch (e) {
        throw new Errors.SetStateClosed.ProgressCreateCallFailed({ progressCode: progress.code }, e);
      }
    } else if (dtoIn.force) {
      try {
        await progressClient.releaseLock();
      } catch (e) {
        if (e.cause?.code !== ProgressConstants.PROGRESS_RELEASE_DOES_NOT_EXIST) {
          throw new Errors.SetStateClosed.ProgressReleaseLockCallFailed({ progressCode: progress.code }, e);
        }
      }

      try {
        await progressClient.setState({ state: "cancelled" });
      } catch (e) {
        DtoBuilder.addWarning(new Warnings.SetStateClosed.ProgressSetStateCallFailed(e.cause?.paramMap));
      }

      try {
        await progressClient.delete();
      } catch (e) {
        if (e.cause?.code !== ProgressConstants.PROGRESS_DELETE_DOES_NOT_EXIST) {
          throw new Errors.SetStateClosed.ProgressDeleteCallFailed({ progressCode: progress.code }, e);
        }
      }

      try {
        progressClient = await ProgressClient.createInstance(config.uuConsoleBaseUri, progress, { session });
      } catch (e) {
        throw new Errors.SetStateClosed.ProgressCreateCallFailed({ progressCode: progress.code }, e);
      }
    }

    try {
      await progressClient.start({
        message: "Progress was started",
        totalWork: <%= context.projectClassName %>Constants.getSetStateClosedStepCount(),
        lockSecret,
      });
    } catch (e) {
      throw new Errors.SetStateClosed.ProgressStartCallFailed({ progressCode: progress.code }, e);
    }

    return progressClient;
  }

  async _runScript(appUri, configuration, lockSecret, session) {
    const scriptEngineClient = new ScriptEngineClient({
      scriptEngineUri: configuration.uuScriptEngineBaseUri,
      consoleUri: configuration.uuConsoleBaseUri,
      consoleCode: <%= context.projectClassName %>Constants.getMainConsoleCode(appUri.getAwid()),
      scriptRepositoryUri: configuration.uuScriptRepositoryBaseUri,
      session,
    });

    const scriptDtoIn = {
      <%= context.namespaceAppCamelCase %>Uri: appUri.toString(),
      lockSecret,
    };

    await scriptEngineClient.runScript({ scriptCode: SCRIPT_CODE, scriptDtoIn });
  }
}

module.exports = new SetStateClosedAbl();
