// uu_appg01_devkit-common/src/templates/__nodejs-app/app/components/configuration.js
const { CacheManager } = require("uu_appg01_cache");
const { Config } = require("uu_appg01_server").Utils;
const { UriBuilder } = require("uu_appg01_server").Uri;
const { UseCaseContext } = require("uu_appg01_server").AppServer;
const { UuAppWorkspace } = require("uu_appg01_server").Workspace;
const { UuTerrClient } = require("uu_territory_clientg01");

const <%= context.projectClassName %>Constants = require("../constants/<%= context.projectClassFile %>-constants");
const <%= context.projectClassName %>UseCaseWarning = require("../api/warnings/<%= context.projectClassFile %>-use-case-warning");
const DtoBuilder = require("./dto-builder");
const ConfigError = require("../api/errors/config-error");

class Configuration {
  static relatedObjectMapCache = CacheManager.getCache(
    `${<%= context.projectClassName %>Constants.UUAPP_CODE}/${<%= context.projectClassName %>Constants.CONFIG_CACHE_KEY}`,
  );

  static async getUuSubAppConfiguration(<%= context.namespaceAppCamelCase %>) {
    return await this.relatedObjectMapCache.get(
      <%= context.namespaceAppCamelCase %>.awid,
      <%= context.projectClassName %>Constants.CONFIG_CACHE_KEY,
      async () => await this.buildConfiguration(<%= context.namespaceAppCamelCase %>),
    );
  }

  static async buildConfiguration(<%= context.namespaceAppCamelCase %>) {
    const mergeConfiguration = (currentConfig, updatedConfig, missingConfigPropertyList) => {
      const configProperties = [...missingConfigPropertyList];
      for (const configProperty of configProperties) {
        if (currentConfig[configProperty] === undefined && updatedConfig[configProperty] !== undefined) {
          currentConfig[configProperty] = updatedConfig[configProperty];
          missingConfigPropertyList.splice(missingConfigPropertyList.indexOf(configProperty), 1);
        }
      }
      return currentConfig;
    };

    let configuration = {};
    let missingConfigPropertyList = ["uuConsoleBaseUri", "uuScriptEngineBaseUri", "uuScriptRepositoryBaseUri"];
    let appClientOpts = {
      baseUri: <%= context.namespaceAppCamelCase %>.uuTerritoryBaseUri,
      session: UseCaseContext.getSession(),
      appUri: UseCaseContext.getUri(),
    };

    configuration = mergeConfiguration(configuration, <%= context.namespaceAppCamelCase %>, missingConfigPropertyList);

    // loads uuSubApp configuration
    if (missingConfigPropertyList.length > 0) {
      let uuAppWorkspace = await UuAppWorkspace.get(<%= context.namespaceAppCamelCase %>.awid);
      configuration = mergeConfiguration(configuration, uuAppWorkspace, missingConfigPropertyList);
    }

    // loads resources from contextProperty configuration
    let <%= context.namespaceAppCamelCase %>CtxProperty;
    let contextPropertyLoadDtoIn = {
      id: <%= context.namespaceAppCamelCase %>.artifactId,
      uuAppMetaModelCode: <%= context.projectClassName %>Constants.UUAPP_CODE,
      propertyCode: <%= context.projectClassName %>Constants.PropertyMap.<%=  context.app.toUpperCase() %>_CONFIG,
    };

    try {
      let contextPropertyLoadDtoOut = await UuTerrClient.ArtifactIfc.ContextProperty.load(
        contextPropertyLoadDtoIn,
        appClientOpts,
      );
      <%= context.namespaceAppCamelCase %>CtxProperty = contextPropertyLoadDtoOut.value;
    } catch (e) {
      DtoBuilder.addWarning(
        new <%= context.projectClassName %>UseCaseWarning("unableToReadConfigurationProperty", e.message, {
          code: contextPropertyLoadDtoIn.propertyCode,
          id: contextPropertyLoadDtoIn.id,
          baseUri: appClientOpts.baseUri,
          uuAppMetaModelCode: contextPropertyLoadDtoIn.uuAppMetaModelCode,
        }),
      );
    }

    if (<%= context.namespaceAppCamelCase %>CtxProperty && missingConfigPropertyList.length > 0) {
      let contextPropertyConfiguration = {
        uuConsoleBaseUri: this._getBaseUri(<%= context.namespaceAppCamelCase %>CtxProperty.uuConsoleUri),
        uuScriptEngineBaseUri: this._getBaseUri(<%= context.namespaceAppCamelCase %>CtxProperty.uuScriptEngineUri),
      };
      configuration = mergeConfiguration(configuration, contextPropertyConfiguration, missingConfigPropertyList);
    }

    // loads resources from ASID configuration
    if (missingConfigPropertyList.length > 0) {
      let asidConfiguration = {
        uuConsoleBaseUri: Config.get("script_console_base_uri"),
        uuScriptEngineBaseUri: Config.get("script_engine_base_uri"),
        uuScriptRepositoryBaseUri: Config.get("script_repository_base_uri"),
      };
      configuration = mergeConfiguration(configuration, asidConfiguration, missingConfigPropertyList);
    }

    // loads resources from uuBT Shared resources
    if (missingConfigPropertyList.length > 0) {
      let artifact;
      try {
        artifact = await UuTerrClient.ArtifactIfc.loadData(
          { id: <%= context.namespaceAppCamelCase %>.artifactId, getSharedResources: true },
          appClientOpts,
        );
      } catch (e) {
        DtoBuilder.addWarning(
          new <%= context.projectClassName %>UseCaseWarning("unableToLoadArtifactForConfiguration", e.message, {
            id: <%= context.namespaceAppCamelCase %>.artifactId,
            baseUri: appClientOpts.baseUri,
            getSharedResources: true,
          }),
        );
      }

      let sharedResourcesMap = artifact?.data?.sharedResourcesMap;
      if (sharedResourcesMap) {
        let sharedResourceConfiguration = {
          uuConsoleBaseUri: this._getBaseUri(sharedResourcesMap[<%= context.projectClassName %>Constants.SharedResources.SCRIPT_CONSOLE]),
          uuScriptEngineBaseUri: this._getBaseUri(sharedResourcesMap[<%= context.projectClassName %>Constants.SharedResources.SCRIPT_ENGINE]),
        };
        configuration = mergeConfiguration(configuration, sharedResourceConfiguration, missingConfigPropertyList);
      }
    }

    configuration.uuAppMetaModelVersion = UuAppWorkspace.getUuAppMetaModelVersion();

    if (missingConfigPropertyList.length > 0) {
      throw new ConfigError.ConfigurationLoadFailed({
        missingKeyList: missingConfigPropertyList,
      });
    } else if (!configuration.uuAppMetaModelVersion) {
      throw new ConfigError.UuAppMetaModelDoesNotExist();
    }

    configuration.uuConsoleCode = <%= context.namespaceAppCamelCase %>CtxProperty?.uuConsoleCode || Config.get("script_console_code");

    configuration.uuAppUuFlsBaseUri = <%= context.namespaceAppCamelCase %>CtxProperty?.uuFlsBaseUri || Config.get("fls_base_uri");
    configuration.uuAppUuSlsBaseUri = <%= context.namespaceAppCamelCase %>CtxProperty?.uuSlsBaseUri || Config.get("sls_base_uri");

    configuration.uuAppBusinessRequestsUri = Config.get("business_request_uri");
    configuration.uuAppBusinessModelUri = Config.get("business_model_uri");
    configuration.uuAppApplicationModelUri = Config.get("application_model_uri");
    configuration.uuAppUserGuideUri = Config.get("user_guide_uri");
    configuration.uuAppWebKitUri = Config.get("web_uri");
    configuration.uuAppProductPortalUri = Config.get("product_portal_uri");

    return configuration;
  }

  static _getBaseUri(uri) {
    if (!uri) {
      return undefined;
    } else {
      return UriBuilder.parse(uri).toUri().getBaseUri().toString();
    }
  }
}

module.exports = Configuration;
