// uu_appg01_devkit-common/src/templates/__nodejs-app/app/components/project-client.js
"use strict";
const { UseCaseContext } = require("uu_appg01_server").AppServer;
const { DaoFactory } = require("uu_appg01_server").ObjectStore;
const { UuTerrClient } = require("uu_territory_clientg01");

const TerritoryConstants = require("../constants/territory-constants");
const DtoBuilder = require("./dto-builder");
const { <%= context.projectClassName %>: Errors } = require("../api/errors/<%= context.projectClassFile %>-error");
const Warnings = require("../api/warnings/<%= context.projectClassFile %>-warning");
const <%=  context.projectClassName %>Constants = require("../constants/<%= context.projectClassFile%>-constants");

class <%= context.projectClassName %>Client {
  constructor(<%= context.namespaceAppCamelCase %>, territoryUri = null, session = null) {
    this.dao = DaoFactory.getDao(<%=  context.projectClassName %>Constants.Schemas.<%=  context.app.toUpperCase() %>_INSTANCE);
    this.<%= context.namespaceAppCamelCase %> = <%= context.namespaceAppCamelCase %>;
    this.uri = UseCaseContext.getUri();
    this.territoryUri = territoryUri ? territoryUri : <%= context.namespaceAppCamelCase %>.uuTerritoryBaseUri;
    this.session = session ? session : UseCaseContext.getSession();
  }

  async createAwsc(location, responsibleRole, permissionMatrix, uuAppMetaModelVersion) {
    const appClientOpts = this.getAppClientOpts();
    const { name, desc } = this.<%= context.namespaceAppCamelCase %>;
    const awscCreateDtoIn = {
      name,
      desc,
      code: `${<%=  context.projectClassName %>Constants.AWSC_PREFIX}/${this.<%= context.namespaceAppCamelCase %>.awid}`,
      location,
      responsibleRole,
      permissionMatrix,
      typeCode: <%=  context.projectClassName %>Constants.UUAPP_CODE,
      uuAppWorkspaceUri: this.uri.getBaseUri(),
      uuAppMetaModelVersion,
    };

    let awsc;
    try {
      awsc = await UuTerrClient.Awsc.create(awscCreateDtoIn, appClientOpts);
    } catch (e) {
      const awscCreateErrorMap = (e.dtoOut && e.dtoOut.uuAppErrorMap) || {};

      const isDup =
        awscCreateErrorMap[TerritoryConstants.AWSC_CREATE_FAILED_CODE] &&
        awscCreateErrorMap[TerritoryConstants.AWSC_CREATE_FAILED_CODE].cause &&
        awscCreateErrorMap[TerritoryConstants.AWSC_CREATE_FAILED_CODE].cause[TerritoryConstants.NOT_UNIQUE_ID_CODE];

      if (isDup) {
        DtoBuilder.addWarning(new Warnings.Init.UuAwscAlreadyCreated());
        awsc = await UuTerrClient.Awsc.get(
          { code: `${<%=  context.projectClassName %>Constants.AWSC_PREFIX}/${this.<%= context.namespaceAppCamelCase %>.awid}` },
          appClientOpts,
        );
      } else {
        DtoBuilder.addUuAppErrorMap(awscCreateErrorMap);
        throw new Errors.CreateAwscFailed(
          { uuTerritoryBaseUri: this.<%= context.namespaceAppCamelCase %>.uuTerritoryBaseUri, awid: this.<%= context.namespaceAppCamelCase %>.awid },
          e,
        );
      }
    }

    this.<%= context.namespaceAppCamelCase %> = await this.dao.updateByAwid({
      awid: this.<%= context.namespaceAppCamelCase %>.awid,
      artifactId: awsc.id,
    });

    return this.<%= context.namespaceAppCamelCase %>;
  }

  async loadAwsc() {
    const appClientOpts = this.getAppClientOpts();

    let awsc;
    try {
      awsc = await UuTerrClient.Awsc.load({ id: this.<%= context.namespaceAppCamelCase %>.artifactId }, appClientOpts);
    } catch (e) {
      throw new Errors.LoadAwscFailed({ artifactId: this.<%= context.namespaceAppCamelCase %>.artifactId }, e);
    }

    return awsc;
  }

  async setAwscState(state) {
    const appClientOpts = this.getAppClientOpts();
    try {
      await UuTerrClient.Awsc.setState(
        {
          id: this.<%= context.namespaceAppCamelCase %>.artifactId,
          state,
        },
        appClientOpts,
      );
    } catch (e) {
      throw new Errors.SetAwscStateFailed({ state, id: this.<%= context.namespaceAppCamelCase %>.artifactId }, e);
    }
  }

  async deleteAwsc() {
    const appClientOpts = this.getAppClientOpts();
    try {
      await UuTerrClient.Awsc.delete({ id: this.<%= context.namespaceAppCamelCase %>.artifactId }, appClientOpts);
    } catch (e) {
      if (e.cause?.code !== TerritoryConstants.ARTIFACT_DOES_NOT_EXIST) {
        throw new Errors.DeleteAwscFailed({ id: this.<%= context.namespaceAppCamelCase %>.artifactId }, e);
      }
    }
  }

  getAppClientOpts() {
    return { baseUri: this.territoryUri, session: this.session, appUri: this.uri };
  }
}

module.exports = <%= context.projectClassName %>Client;
