// uu_appg01_devkit-common/src/templates/__nodejs-uuscriptlib/src/project-client.js
const AppClient = require("uu_appg01_server").AppClient;
const { UseCaseError } = require("uu_appg01_server").AppServer;

const { session, dtoOut } = scriptContext;

/*@@viewOn:names*/
const Names = {
  SCRIPT_LIB_NAME: "<%= context.nameWithVendor %>",
  CLASS_NAME: "<%=  context.projectClassName %>Client",
};
/*@@viewOff:names*/

/*@@viewOn:errors*/
const Errors = {
  ERROR_PREFIX: `${Names.SCRIPT_LIB_NAME}/${Names.CLASS_NAME}/`,

  Load<%=  context.namespaceApp %>Failed: class extends UseCaseError {
    constructor(paramMap, cause) {
      super({ dtoOut, paramMap, status: 400 }, cause);
      this.message = "Calling sys/uuAppWorkspace/load failed.";
      this.code = `${Errors.ERROR_PREFIX}load<%=  context.namespaceApp %>Failed`;
    }
  },

  Get<%=  context.namespaceApp %>Failed: class extends UseCaseError {
    constructor(paramMap, cause) {
      super({ dtoOut, paramMap, status: 400 }, cause);
      this.message = "Calling sys/uuAppWorkspace/get failed.";
      this.code = `${Errors.ERROR_PREFIX}get<%=  context.namespaceApp %>Failed`;
    }
  },

  InitFinalizeFailed: class extends UseCaseError {
    constructor(paramMap, cause) {
      super({ dtoOut, paramMap, status: 400 }, cause);
      this.message = "Calling sys/uuAppWorkspace/_initFinalize failed.";
      this.code = `${Errors.ERROR_PREFIX}initFinalizeFailed`;
    }
  },

  InitFinalizeRollbackFailed: class extends UseCaseError {
    constructor(paramMap, cause) {
      super({ dtoOut, paramMap, status: 400 }, cause);
      this.message = "Calling sys/uuAppWorkspace/_initFinalizeRollback failed.";
      this.code = `${Errors.ERROR_PREFIX}initFinalizeRollbackFailed`;
    }
  },

  SetStateClosedFinalizeFailed: class extends UseCaseError {
    constructor(paramMap, cause) {
      super({ dtoOut, paramMap, status: 400 }, cause);
      this.message = "Calling sys/uuAppWorkspace/_setStateClosedFinalize failed.";
      this.code = `${Errors.ERROR_PREFIX}setStateClosedFinalizeFailed`;
    }
  },

  ClearFinalizeFailed: class extends UseCaseError {
    constructor(paramMap, cause) {
      super({ dtoOut, paramMap, status: 400 }, cause);
      this.message = "Calling sys/uuAppWorkspace/_clearFinalize failed.";
      this.code = `${Errors.ERROR_PREFIX}clearFinalizeFailed`;
    }
  },
};

/*@@viewOff:errors*/

class <%=  context.projectClassName %>Client {
  constructor(baseUri) {
    this.appClient = new AppClient({ baseUri, session });
    // base uri can be used as parameter in error
    this.baseUri = baseUri;
  }

  async load() {
    let <%=  context.app %>;
    try {
      <%=  context.app %> = await this.appClient.cmdGet("sys/uuAppWorkspace/load");
    } catch (e) {
      throw new Errors.Load<%=  context.namespaceApp %>Failed({ baseUri: this.baseUri }, e);
    }
    return <%=  context.app %>;
  }

  async get() {
    let <%=  context.app %>;
    try {
      <%=  context.app %> = await this.appClient.cmdGet("sys/uuAppWorkspace/get");
    } catch (e) {
      throw new Errors.Get<%=  context.namespaceApp %>Failed({ baseUri: this.baseUri }, e);
    }
    return <%=  context.app %>;
  }

  async initFinalize(lockSecret) {
    let <%=  context.app %>;
    try {
      <%=  context.app %> = await this.appClient.cmdPost("sys/uuAppWorkspace/_initFinalize", { lockSecret });
    } catch (e) {
      throw new Errors.InitFinalizeFailed({ baseUri: this.baseUri }, e);
    }
    return <%=  context.app %>;
  }

  async initFinalizeRollback(lockSecret) {
    let <%=  context.app %>;
    try {
      <%=  context.app %> = await this.appClient.cmdPost("sys/uuAppWorkspace/_initFinalizeRollback", { lockSecret });
    } catch (e) {
      throw new Errors.InitFinalizeRollbackFailed({ baseUri: this.baseUri }, e);
    }
    return <%=  context.app %>;
  }

  async setStateClosedFinalize(lockSecret) {
    let <%=  context.app %>;
    try {
      <%=  context.app %> = await this.appClient.cmdPost("sys/uuAppWorkspace/_setStateClosedFinalize", { lockSecret });
    } catch (e) {
      throw new Errors.SetStateClosedFinalizeFailed({ baseUri: this.baseUri }, e);
    }
    return <%=  context.app %>;
  }

  async clearFinalize(lockSecret) {
    let <%=  context.app %>;
    try {
      <%=  context.app %> = await this.appClient.cmdPost("sys/uuAppWorkspace/_clearFinalize", { lockSecret });
    } catch (e) {
      throw new Errors.ClearFinalizeFailed({ baseUri: this.baseUri }, e);
    }
    return <%=  context.app %>;
  }
}

module.exports = <%=  context.projectClassName %>Client;
